File: src/assistant/events.py
"""AG2 Assistant application events that ride the AG2 stream.

These are ordinary AG2 events (``BaseEvent`` subclasses), so they serialize,
persist (via ``EventLogWriter``), replay, and stream over the wire with exactly
the same `{type, data}` shape as every native event — no parallel model. The GUI
renders them by type alongside AG2's own events.

We only define what AG2 main doesn't already cover. Conversation, tool, voice,
HITL (`HumanInputRequest`), and task-lifecycle (`TaskStarted`/`TaskCompleted`/…)
events are AG2 native and reused as-is. These add the AG2 Assistant-specific signals:
a task was spawned from a chat, a deliverable was produced, a task was scheduled,
and the durable-inquiry lifecycle.
"""

from ag2.events import BaseEvent, Field


class AssistantEvent(BaseEvent):
    """Marker base for AG2 Assistant app-level events (no fields of its own)."""


class TaskCreated(AssistantEvent):
    """A background task was spawned (e.g. from a chat) — renders as a task card."""

    task_id: str = Field(kw_only=False)  # type: ignore[assignment]
    title: str = ""
    kind: str = "task"  # "task" | "scheduled"


class TaskScheduled(AssistantEvent):
    """A task was scheduled (or rescheduled) to run later — renders as a schedule note."""

    task_id: str = Field(kw_only=False)  # type: ignore[assignment]
    scheduled_for: str = ""  # ISO 8601
    recurrence: str = ""  # 5-field cron; "" for one-off
    recurrence_desc: str = ""  # human-readable cron description ("" for one-off)


class DeliverableProduced(AssistantEvent):
    """A task produced a deliverable — renders as a deliverable item with a preview."""

    task_id: str = Field(kw_only=False)  # type: ignore[assignment]
    deliverable_id: str = ""
    description: str = ""
    preview: str = ""  # short preview; full asset fetched via REST
    path: str = ""  # workspace-relative file the deliverable was saved as ("" if none)


class Attachment(AssistantEvent):
    """A file the user attached to a chat turn, saved to the workspace — renders in
    the thread as a thumbnail (images) or a file chip. `path` is workspace-relative."""

    path: str = Field(kw_only=False)  # type: ignore[assignment]
    name: str = ""
    media_type: str = ""


class ImageGenerated(AssistantEvent):
    """An image was generated/edited and saved to the workspace — renders inline as
    a clickable thumbnail. `path` is workspace-relative (served via /api/files/raw)."""

    path: str = Field(kw_only=False)  # type: ignore[assignment]
    prompt: str = ""
    media_type: str = "image/png"


class A2UISurface(AssistantEvent):
    """A declarative A2UI surface for the web client to render inline.

    This is the assistant app's stable envelope around upstream A2UI messages:
    once ``autogen.beta.a2ui`` is available locally, validated A2UI message
    events can be projected into this shape without changing the Svelte renderer.
    """

    surface_id: str = Field(kw_only=False)  # type: ignore[assignment]
    catalog_id: str = "https://ag2.ai/assistant/a2ui/catalog.json"
    version: str = "v1.0"
    component: dict = Field(default_factory=dict)  # type: ignore[assignment]
    data: dict = Field(default_factory=dict)  # type: ignore[assignment]
    title: str = ""
    intent: str = ""


class A2UISurfaceDataUpdated(AssistantEvent):
    """Durable data-model snapshot emitted after an A2UI server action."""

    surface_id: str = Field(kw_only=False)  # type: ignore[assignment]
    data: dict = Field(default_factory=dict)  # type: ignore[assignment]


class A2UIActionSubmitted(AssistantEvent):
    """A user action was accepted by the backend and is being handled."""

    surface_id: str = Field(kw_only=False)  # type: ignore[assignment]
    action_name: str = ""


class SubagentTrace(AssistantEvent):
    """One inner event from a subagent's own run, forwarded onto the parent task
    stream so the GUI nests it under the subagent card — live, persistent (rides
    the task event log), and recursive: a nested subagent's events arrive as
    further SubagentTraces and nest one level deeper.

    `subagent_id` matches the subagent's TaskStarted/TaskCompleted `task_id` (the
    card key); `inner` is the wrapped event in wire shape ``{type, data}``.
    """

    subagent_id: str = Field(kw_only=False)  # type: ignore[assignment]
    inner: dict = Field(default_factory=dict)  # type: ignore[assignment]


class InquiryRaised(AssistantEvent):
    """A durable HITL question/permission was raised (backed by InquiryStore).

    Mirrors AG2's transient ``HumanInputRequest`` but is durable and task-scoped,
    so it survives restarts and is answerable from any surface.
    """

    inquiry_id: str = Field(kw_only=False)  # type: ignore[assignment]
    task_id: str = ""
    question: str = ""
    detail: str = ""  # secondary context, e.g. the exact code/command to be run
    options: list[str] = Field(default_factory=list)  # type: ignore[assignment]
    kind: str = "question"  # "question" | "permission"


class InquiryAnswered(AssistantEvent):
    """A durable inquiry reached a terminal state — resolves the matching HITL item.

    Emitted for every resolution, not just a real answer: ``status`` is one of
    "answered" | "expired" (timed out unanswered) | "cancelled" (owning task ended)
    so the GUI can retire a stranded permission/question card and say *how* it
    resolved instead of leaving live-looking buttons that do nothing. ``answer`` is
    the chosen text for "answered", empty otherwise.
    """

    inquiry_id: str = Field(kw_only=False)  # type: ignore[assignment]
    answer: str = ""
    status: str = "answered"


class FeedbackGiven(AssistantEvent):
    """The user reacted to a generated item (👍/👎) with a mandatory reason. Rides the
    stream so the thumb state persists + replays (and shows in the AG2 inspector); a
    learner agent distils it into the memory profile.

    `target_kind` is "message" | "image" | "deliverable"; `target_id` is that kind's
    stable key (message → the reply's `created_at`; image → workspace path; deliverable
    → deliverable_id). `content` is an excerpt of what was rated and `request` the user's
    intent that produced it — both passed to the learner so it generalises correctly.
    """

    target_id: str = Field(kw_only=False)  # type: ignore[assignment]
    target_kind: str = "message"  # "message" | "image" | "deliverable"
    sentiment: str = "up"  # "up" | "down"
    reason: str = ""
    content: str = ""  # excerpt of the rated output
    request: str = ""  # the user's ask / task objective that produced it


class FeedbackCleared(AssistantEvent):
    """The user retracted a rating (toggled the 👍/👎 off). Rides the stream so the
    cleared state persists + replays — the GUI projects the thumb back to neutral.

    Carries no sentiment or reason: unmarking takes back only the *visible* rating.
    Any preference the learner already distilled into the memory profile is left
    untouched (editable directly in Settings → Memory), so this event never runs the
    learner. `target_kind`/`target_id` match the original FeedbackGiven's key.
    """

    target_id: str = Field(kw_only=False)  # type: ignore[assignment]
    target_kind: str = "message"  # "message" | "image" | "deliverable"


class TurnCancelled(AssistantEvent):
    """The user stopped a turn while it was running. Emitted after AG2 cancels the
    run, so it lands after whatever the turn had already produced (tool calls and
    their results stay in the history) — the thread shows where the work stopped,
    and replays that way."""

    chat_id: str = Field(kw_only=False)  # type: ignore[assignment]
    reason: str = "Stopped"


class TurnFailed(AssistantEvent):
    """A turn ended in an error (timeout, provider fault, connection drop) rather
    than a reply. Emitted before the turn's history is persisted, so it lands after
    whatever the turn had already produced — the thread keeps that work and says why
    it stopped instead of simply ending mid-air.

    `error` is a short, user-facing sentence; the traceback stays in the debug record
    written by ``capture_failure``, never in the chat."""

    chat_id: str = Field(kw_only=False)  # type: ignore[assignment]
    error: str = ""

File: web/src/schemas/events.ts
// AG2 wire events. `type` arrives fully qualified (….ModelResponse); project.ts
// narrows on its tail. Each known event pins the fields the reducer reads; an
// unknown type falls through to the catch-all so a new backend event never
// breaks the UI.
import { z } from 'zod'
import type { KeyedToolCard as ToolCard } from '../lib/toolcards.ts'
import type { A2UIComponent } from '../lib/a2ui.ts'

const TextPart = z.object({ __event__: z.string().optional(), content: z.string().optional() })

// The `data` shape per event the reducer folds. project.ts parses with these
// AFTER narrowing on tail(type) — WireEvent itself never rejects a type.
export const EventData = {
  ModelRequest: z.object({ parts: z.array(TextPart).optional() }),
  ModelMessageChunk: z.object({ content: z.string().optional() }),
  ModelResponse: z.object({
    message: z.object({ content: z.string().optional() }).nullish(),
    tool_calls: z.object({ calls: z.array(z.unknown()).optional() }).nullish(),
  }),
  ToolCallsEvent: z.object({
    calls: z.array(z.object({ name: z.string(), arguments: z.unknown().optional() })).optional(),
  }),
  TaskCreated: z.object({
    task_id: z.string(),
    title: z.string().optional(),
    kind: z.string().optional(),
  }),
  TaskScheduled: z.object({
    scheduled_for: z.string().optional(),
    recurrence: z.string().optional(),
    recurrence_desc: z.string().optional(),
  }),
  // Shared by TaskStarted / TaskCompleted / TaskFailed / TaskCancelled. A root
  // task event has no agent_name; anything else folds into a subagent card, which
  // is why agent_name and objective matter here.
  TaskLifecycle: z.object({
    agent_name: z.string().optional(),
    task_id: z.string().optional(),
    objective: z.string().optional(),
    result: z.unknown().optional(),
    error: z.unknown().optional(),
    reason: z.string().optional(),
  }),
  SubagentTrace: z.object({
    subagent_id: z.string().optional(),
    inner: z.looseObject({ type: z.string() }).nullish(),
  }),
  ObserverAlert: z.object({ message: z.string() }),
  DeliverableProduced: z.object({
    task_id: z.string().optional(),
    deliverable_id: z.string(),
    description: z.string().optional(),
    preview: z.string().optional(),
    path: z.string().optional(),
  }),
  ImageGenerated: z.object({ path: z.string(), prompt: z.string().optional() }),
  A2UISurface: z.object({
    surface_id: z.string(),
    version: z.string().optional(),
    catalog_id: z.string().optional(),
    title: z.string().optional(),
    intent: z.string().optional(),
    component: z.record(z.string(), z.unknown()).optional(),
    components: z.array(z.unknown()).optional(),
    data: z.record(z.string(), z.unknown()).optional(),
  }),
  A2UISurfaceDataUpdated: z.object({
    surface_id: z.string(),
    data: z.record(z.string(), z.unknown()).optional(),
  }),
  A2UIActionSubmitted: z.object({ surface_id: z.string() }),
  // ag2 A2UIMessageEvent carries exactly one canonical A2UI message dict.
  A2UIMessageEvent: z.object({ message: z.record(z.string(), z.unknown()) }),
  Attachment: z.object({ path: z.string(), name: z.string().optional() }),
  InquiryRaised: z.object({
    inquiry_id: z.string(),
    question: z.string(),
    detail: z.string().optional(),
    options: z.array(z.string()).optional(),
    kind: z.string().optional(),
  }),
  InquiryAnswered: z.object({
    inquiry_id: z.string(),
    answer: z.string().optional(),
    status: z.string().optional(),
  }),
  TurnCancelled: z.object({ reason: z.string().optional() }),
  TurnFailed: z.object({ error: z.string().optional() }),
  // FeedbackGiven and FeedbackCleared share this payload; only the clear ignores
  // sentiment. target_kind picks which item key target_id matches on.
  FeedbackGiven: z.object({
    target_kind: z.enum(['message', 'image', 'deliverable']),
    target_id: z.string(),
    sentiment: z.string().optional(),
    reason: z.string().optional(),
  }),
} as const

// gateway/wire.py to_wire sends exactly {type, data} — nothing else.
export const WireEvent = z.object({
  type: z.string(),
  data: z.record(z.string(), z.unknown()).default({}),
})
export type WireEvent = z.infer<typeof WireEvent>

// Every event's data carries its production time (ag2 BaseEvent.created_at, Unix
// seconds); the reducer stamps the items an event produces with it.
export const EventMeta = z.object({ created_at: z.number().optional() })
export type EventMeta = z.infer<typeof EventMeta>

// The event names project.ts folds; anything else is ignored by design.
export const HANDLED_EVENTS = [
  'ModelRequest', 'DrainedModelRequest', 'ModelMessageChunk', 'ModelResponse',
  'ToolCallsEvent', 'TaskCreated', 'TaskScheduled', 'TaskStarted', 'TaskCompleted',
  'TaskFailed', 'TaskCancelled', 'SubagentTrace', 'ObserverAlert', 'DeliverableProduced',
  'ImageGenerated', 'A2UISurface', 'A2UISurfaceDataUpdated', 'A2UIActionSubmitted',
  'A2UIMessageEvent', 'Attachment', 'InquiryRaised', 'InquiryAnswered', 'TurnCancelled',
  'TurnFailed', 'FeedbackGiven', 'FeedbackCleared',
] as const
export type HandledEvent = (typeof HANDLED_EVENTS)[number]

// Frames the stream socket receives. stream_bridge.py stamps `chat` (the stream
// id) on ready/turn_end/error; the failure text rides in `message`, not `error`.
export const EventFrame = z.object({ event: WireEvent })
export type EventFrame = z.infer<typeof EventFrame>

export const ReadyFrame = z.object({ type: z.literal('ready'), chat: z.string().optional() })
export type ReadyFrame = z.infer<typeof ReadyFrame>

// `role` comes from the voice socket, which reuses this frame (app.py:3565).
export const TurnEndFrame = z.object({
  type: z.literal('turn_end'),
  chat: z.string().optional(),
  role: z.string().optional(),
})
export type TurnEndFrame = z.infer<typeof TurnEndFrame>

export const QueuedFrame = z.object({
  type: z.literal('queued'),
  text: z.string().optional(),
  chat: z.string().optional(),
})
export type QueuedFrame = z.infer<typeof QueuedFrame>

export const ErrorFrame = z.object({
  type: z.literal('error'),
  message: z.string().optional(),
  chat: z.string().optional(),
})
export type ErrorFrame = z.infer<typeof ErrorFrame>

export const ControlFrame = z.discriminatedUnion('type', [
  ReadyFrame,
  TurnEndFrame,
  QueuedFrame,
  ErrorFrame,
])
export type ControlFrame = z.infer<typeof ControlFrame>

export const ServerFrame = z.union([EventFrame, ControlFrame])
export type ServerFrame = z.infer<typeof ServerFrame>

// Frames the voice socket receives on top of the shared event frame: the same
// ready/turn_end/error plus its own live transcript (app.py:3565-3616).
export const TranscriptFrame = z.object({
  type: z.literal('transcript'),
  role: z.enum(['user', 'agent']),
  text: z.string(),
  final: z.boolean().optional(),
})
export type TranscriptFrame = z.infer<typeof TranscriptFrame>

export const VoiceFrame = z.union([
  EventFrame,
  z.discriminatedUnion('type', [ReadyFrame, TranscriptFrame, TurnEndFrame, ErrorFrame]),
])
export type VoiceFrame = z.infer<typeof VoiceFrame>

// One attachment a turn frame carries: the composer base64-encodes the file and the
// gateway decodes it into an AG2 input (app.py:3689 _decode_attachments).
export type AttachmentPayload = { name: string; mime: string; data: string }

// Frames the client sends. A turn is the bare {text, attachments} shape.
export type ClientFrame =
  // `model` rides only the message that CREATES a chat: a Text model chosen
  // before the chat existed, adopted as its override (ADR 0025).
  | { text: string; attachments?: AttachmentPayload[]; model?: string }
  | { type: 'cancel' }
  | { type: 'answer'; id: string; answer: string }
  | { type: 'feedback'; [k: string]: unknown }
  | { type: 'feedback_clear'; [k: string]: unknown }
  | { type: 'a2ui'; message: unknown }

// Thread items — the projection project.ts folds events into. These never cross
// the wire, so they are plain types rather than schemas.
export type Feedback = { sentiment?: string; reason?: string } | null

// `id` keys the rendered list; every producer takes it from lib/ids.ts so items
// from the reducer, the controller and a2ui.js never collide. `at` is the source
// event's created_at (Unix seconds).
type ItemBase = { id: number; at?: number }

// One card per tool call the thread renders alongside the chips — the union
// lib/toolcards.ts builds, keyed by `id` for the rendered list.
export type { ToolCard }

export type ThreadItem =
  // `voice` marks a bubble spoken in a live voice session — UserMessage/AgentMessage style it.
  | (ItemBase & { kind: 'user'; text: string; queued?: boolean; voice?: boolean })
  | (ItemBase & { kind: 'agent'; text: string; streaming?: boolean; empty?: boolean; voice?: boolean; feedback?: Feedback })
  | (ItemBase & { kind: 'tools'; names: { name: string; n: number }[]; cards: ToolCard[] })
  | (ItemBase & { kind: 'note'; icon: string; text: string; alert?: boolean; ends?: boolean; pending?: boolean; a2uiActionPending?: boolean; surfaceId?: string })
  | (ItemBase & { kind: 'taskcard'; taskId: string; title?: string; scheduled: boolean })
  | (ItemBase & { kind: 'subagent'; taskId: string; agent: string; objective: string; status?: string; result?: string; error?: string; items: ThreadItem[] })
  | (ItemBase & { kind: 'deliverable'; taskId?: string; deliverableId: string; description?: string; preview?: string; path?: string; feedback?: Feedback })
  | (ItemBase & { kind: 'genimage'; path: string; prompt?: string; feedback?: Feedback })
  | (ItemBase & { kind: 'attachment'; path: string; name?: string })
  | (ItemBase & { kind: 'inquiry'; inquiryId: string; question: string; detail: string; options: string[]; qkind?: string; resolved: boolean; answer?: string; resolution?: string })
  | (ItemBase & { kind: 'a2ui'; surfaceId: string; version?: string; catalogId?: string; title?: string; intent?: string; component: A2UIComponent; components?: A2UIComponent[]; data: Record<string, unknown>; messages?: unknown[] })
